tests/network/services.go

```diff
@@ -247,67 +235,61 @@ var _ = SIGDescribe("[Serial]Services", func() {
 		})
 
 		Context("with a service matching the vmi exposed", func() {
-			var jobsToCleanup []func() error
-			var serviceManager *vmiServiceManager
-			var serviceNamePrefix string
-
-			BeforeEach(func() {
-				serviceNamePrefix = "myservice"
-				serviceManager = newVMIServiceManager(inboundVMI, servicePort, serviceNamePrefix, selectorLabelKey, selectorLabelValue)
-
-				services := serviceManager.buildK8sServicesSpec()
-				Expect(services).NotTo(BeEmpty(), "a service should be exposed per each iface. At least one *must* be present.")
-
-				for _, exposedService := range services {
-					_, err := virtClient.CoreV1().Services(inboundVMI.Namespace).Create(&exposedService)
-					Expect(err).NotTo(HaveOccurred())
-				}
-
-				// reset the job cleanup functions list
-				jobsToCleanup = nil
-			})
+			var jobToCleanup func() error
+			var service *k8sv1.Service
 
 			AfterEach(func() {
-				var errors []error
-				for _, exposedService := range serviceManager.services {
-					err := virtClient.CoreV1().Services(inboundVMI.Namespace).Delete(exposedService.Name, &k8smetav1.DeleteOptions{})
-					errors = append(errors, err)
-				}
-				for _, err := range errors {
-					Expect(err).NotTo(HaveOccurred())
-				}
+				Expect(jobToCleanup).NotTo(BeNil(), "a k8sv1.Job cleaning up function should exist")
+				Expect(jobToCleanup()).To(Succeed(), "cleaning up the k8sv1.Job entity should have succeeded.")
+				jobToCleanup = nil
 			})
 
 			AfterEach(func() {
-				cleanupJobs(jobsToCleanup)
+				Expect(cleanupService(inboundVMI.GetNamespace(), service.Name)).To(Succeed(), "cleaning up the k8sv1.Service entity should have succeeded.")
 			})
 
-			It("should be able to reach the vmi based on labels specified on the vmi", func() {
-				for _, exposedService := range serviceManager.services {
-					jobCleanupFunc := assertConnectivityToService(exposedService.Name, inboundVMI.Namespace, servicePort)
-					jobsToCleanup = append(jobsToCleanup, jobCleanupFunc)
+			table.DescribeTable("should be able to reach the vmi based on labels specified on the vmi", func(ipFamily k8sv1.IPFamily) {
+				serviceName := "myservice"
+				if ipFamily == k8sv1.IPv6Protocol {
+					isDualStackCluster, _ := libnet.IsClusterDualStack(virtClient)
+					if !isDualStackCluster {
+						Skip("IPv6 tests required a dual stack cluster.")
+					}
+
+					serviceName = serviceName + "v6"
+					service = buildIPv6ServiceSpec(serviceName, servicePort, servicePort, selectorLabelKey, selectorLabelValue)
+				} else {
+					service = buildServiceSpec(serviceName, servicePort, servicePort, selectorLabelKey, selectorLabelValue)
 				}
+
+				_, err := virtClient.CoreV1().Services(inboundVMI.Namespace).Create(service)
+				Expect(err).NotTo(HaveOccurred(), "the k8sv1.Service entity should have been created.")
```

Comment:
We cannot use BeforeEach, but we should still differentiate between the test and the setup of resources. Could you enclose the setup phase in a `By` callback?

```
By("blah", func() {
// setup resources
})
```

One issue it has is that you can't use variables set by the block. So you would need to initialize those variables before `By`. If you would not be down to live with those variables, please at least add `By` for setup and for the test execution.